File: test/artifacts.cpp
#include <cstdio>
#include <vector>
#include <deque>
#include <algorithm>
#include <climits>
using namespace std;

int main() {
    int n, m;
    scanf("%d %d", &n, &m);

    vector<pair<int,int>> items; // (value, epoch)
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++) {
            int a;
            scanf("%d", &a);
            items.push_back({a, i});
        }

    sort(items.begin(), items.end());

    int total = n * m;
    vector<deque<int>> eq(n); // values per epoch in window
    int covered = 0;
    long long cur_sum = 0;
    long long best_range = LLONG_MAX, best_sum = LLONG_MAX;
    int bl = -1, br = -1;

    for (int l = 0, r = 0; r < total; r++) {
        auto [v, ep] = items[r];
        if (eq[ep].empty()) { covered++; cur_sum += v; }
        eq[ep].push_back(v);

        while (covered == n) {
            long long rng = (long long)items[r].first - items[l].first;
            if (rng < best_range || (rng == best_range && cur_sum < best_sum)) {
                best_range = rng;
                best_sum = cur_sum;
                bl = l; br = r;
            }
            auto [lv, le] = items[l];
            eq[le].pop_front();
            if (eq[le].empty()) { covered--; cur_sum -= lv; }
            else cur_sum += eq[le].front() - lv;
            l++;
        }
    }

    // Reconstruct: pick first (minimum) occurrence per epoch in [bl, br]
    vector<int> sel(n, -1);
    for (int i = bl; i <= br; i++) {
        auto [v, ep] = items[i];
        if (sel[ep] == -1) sel[ep] = v;
    }

    sort(sel.begin(), sel.end());
    for (int i = 0; i < n; i++)
        printf("%d%c", sel[i], " \n"[i == n - 1]);
}

File: test/artifacts.py
import sys
from collections import deque

input = sys.stdin.readline

n, m = map(int, input().split())
items = []
for i in range(n):
    for v in map(int, input().split()):
        items.append((v, i))

items.sort()

eq = [deque() for _ in range(n)]
covered = 0
cur_sum = 0
best_range = float('inf')
best_sum = float('inf')
bl = br = 0
l = 0

for r in range(n * m):
    v, ep = items[r]
    if not eq[ep]:
        covered += 1
        cur_sum += v
    eq[ep].append(v)

    while covered == n:
        rng = items[r][0] - items[l][0]
        if rng < best_range or (rng == best_range and cur_sum < best_sum):
            best_range = rng
            best_sum = cur_sum
            bl, br = l, r

        lv, le = items[l]
        eq[le].popleft()
        if not eq[le]:
            covered -= 1
            cur_sum -= lv
        else:
            cur_sum += eq[le][0] - lv
        l += 1

sel = [0] * n
seen = [False] * n
for i in range(bl, br + 1):
    v, ep = items[i]
    if not seen[ep]:
        seen[ep] = True
        sel[ep] = v

sel.sort()
print(*sel)
